### ISSUE TITLE
Spherical harmonic incongruences

### ISSUE THREAD

**Issue by arokem** [scipy/scipy#4887]
In the course of our work on a neuroimaging application using spherical harmonics (over here: https://github.com/nipy/dipy), my confederates and I stumbled on some incongruences in the calculation of spherical harmonics. 

The upshot is that we compared the scipy implementation to our own implementation (which you can see in the notebook referred below) and to the one in sympy.mpmath. All three implementations diverge by some small number from each other (order 10e-15 to 10e-16), but the differences between scipy and the others is an order of magnitude larger. 

A notebook demonstrating this (including the code of our implementation) is here: 

http://nbviewer.ipython.org/gist/arokem/d9926589836aafa356d2

The conversation that led to this is here:

https://github.com/nipy/dipy/pull/612

The differences are rather small, but exacerbated in our hands, due to multiplication (that's the test at the bottom of the notebook), and I don't know what to make of them. I was hoping someone here would have ideas as to where/why they arise. 

Thanks!

**Comment by argriffing** [comment]
> my confederates and I stumbled on some incongruences in the calculation of spherical harmonics.

The scipy implementation of `sph_harm` was recently changed, and it passed the scipy unit tests.  Is there a single input/output pair that best shows the incongruence?

The recent change in the scipy implementation:
https://github.com/scipy/scipy/commit/5a8ca15273f08801ce712ee89a2ad1731ae51789.

The scipy unit tests:
https://github.com/scipy/scipy/blob/maintenance/0.16.x/scipy/special/tests/test_basic.py#L2892

**Comment by argriffing** [comment]
Part of the formula for the spherical harmonic function is like `(n-m)! / (n+m)!`.  The dipy implementation looks like it deals with this well through differences of log gamma, whereas the scipy implementation does something questionable with `poch`.
```
>>> n = 16
>>> m = 16
>>> poch(n+m+1, -2*m)
3.8003907548547441e-36
```

The scipy unit tests only cover small (m, n) pairs like
0, 0
-2, 2
2, 2
2, 4
4, 4

**Comment by pv** [comment]
Well, the result from poch is correct at least for those parameters.

**Comment by argriffing** [comment]
> Well, the result from poch is correct at least for those parameters.

I agree it is close, but I think it would explain differences like "order 10e-15 to 10e-16".  Should this issue be closed?  It seems pretty precise to me, for double precision.

```python
>>> n = 16
>>> m = 16
>>> np.exp(0.5 * (gammaln(n-m+1) - gammaln(n+m+1)))
1.9494590928908398e-18
>>> np.sqrt(poch(n+m+1, -2*m))
1.9494590928908317e-18
```

> The differences are rather small, but exacerbated in our hands, due to multiplication

Maybe this could be avoided by working in log space or something?

**Comment by pv** [comment]
@argriffing: we do test orders up to 100 (though randomly sampled), cf.
test_mpmath.py. Absolute accuracy of 1e-15...1e-16 is about as good as
you can get in floating point. (If a calculation is such that changes of
this order cause large differences, the calculation is just numerically
unstable. The solution is usually to do something different :)
.
However, there is a cause for loss accuracy in spherical harmonic
evaluation, which comes from the use of the Legendre polynomial formula.
Basically, `x=cos(theta) ~ 1 - theta**2/2` close to theta=0 loses half
of the decimals in accuracy for m=1, which has P^m_l(x) \propto sqrt(1 -
x^2)`. This should be special cased (and the mpmath test accuracy bumped
up). IIRC for other values of m the zero at x=1 is at least linear, so
you end up with the standard machine epsilon accuracy.
.
The dipy implementation quoted in the links however has exactly the same
flaw, so that doesn't seem to be the cause here.

**Comment by pv** [comment]
@argriffing: poch is probably more accurate than gammaln. However, as
you say, relative differences of machine epsilon size can quite easily
come from rounding errors.

**Comment by argriffing** [comment]
> `print(np.max((model_full.B_reg - expected * B_reg)))`

I wonder whether the regression is on the left hand side vs. the right hand side of this comparison.  I think both sides involve the calculation of `sph_harm`, and the right-hand side seems to not differ significantly from the value that mpmath calculates.

**Comment by argriffing** [comment]
I looked at the `dipy` code, and now I can confirm this looks like a scipy bug.  I think the issue is that scipy is mistakenly creating an output array of dtype `complex64` instead of dtype `complex128`.

```python
>>> import numpy as np
>>> from scipy.special import sph_harm
>>> m = np.array(0, dtype=np.int64)
>>> n = np.array([0, 1], dtype=np.int64)
>>> v = sph_harm(m, n, 0, 0)
>>> v
array([ 0.28209481+0.j,  0.48860252+0.j], dtype=complex64)
```

**Comment by argriffing** [comment]
A simplified example:
```python
>>> sph_harm(0, 0, 0, 0)
(0.28209479177387814+0j)
>>> sph_harm(0, [0], 0, 0)
array([ 0.28209481+0.j], dtype=complex64)
```

**Comment by rgommers** [comment]
Is this fixed in master now after gh-4896?

**Comment by argriffing** [comment]
The simplified example above is now fixed in the scipy development branch.

```python
>>> from scipy.special import sph_harm
>>> sph_harm(0, 0, 0, 0)
(0.28209479177387814+0j)
>>> sph_harm(0, [0.], 0, 0)
array([ 0.28209479+0.j])
>>> sph_harm(0, [0], 0, 0)
array([ 0.28209479+0.j])
>>> import scipy
>>> scipy.__version__
'0.17.0.dev0+f3df41b'
```

I'll wait a bit for @arokem to verify before closing.

**Comment by arokem** [comment]
Yes - I just verified that with the original use-case. Thanks!
